src/chainstate/stacks/boot/mod.rs

```diff
@@ -0,0 +1,3428 @@
+ (at your option) any later version.
+
+ Blockstack is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY, including without the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use chainstate::stacks::StacksAddress;
+use chainstate::stacks::db::StacksChainState;
+use chainstate::stacks::Error;
+use chainstate::stacks::StacksBlockHeader;
+
+use burnchains::bitcoin::address::BitcoinAddress;
+use burnchains::Address;
+use address::AddressHashMode;
+
+use chainstate::burn::db::sortdb::SortitionDB;
+
+use vm::types::{
+    StandardPrincipalData,
+    QualifiedContractIdentifier,
+    Value,
+    TupleData,
+    PrincipalData,
+    SequenceData
+};
+
+use chainstate::stacks::StacksBlockId;
+
+use burnchains::Burnchain;
+
+use vm::representations::ContractName;
+
+use util::hash::Hash160;
+
+use std::convert::TryFrom;
+use std::convert::TryInto;
+use std::boxed::Box;
+
+pub const STACKS_BOOT_CODE_CONTRACT_ADDRESS : &'static str = "ST000000000000000000002AMW42H";
+
+pub const STACKS_BOOT_CODE : &'static [(&'static str, &'static str)] = &[
+    ("pox-api", std::include_str!("pox.clar")),     // Clarity prevents us from using anything shorter than 5 characters, so 'pox' isn't an option :(
+    ("lookup", std::include_str!("lockup.clar"))
+];
+
+pub fn boot_code_addr() -> StacksAddress {
+    StacksAddress::from_string(&STACKS_BOOT_CODE_CONTRACT_ADDRESS.clone()).unwrap()
+}    
+
+pub fn boot_code_id(name: &str) -> QualifiedContractIdentifier {
+    QualifiedContractIdentifier::new(StandardPrincipalData::from(boot_code_addr()), ContractName::try_from(name.to_string()).unwrap())
+}
+
+pub fn make_contract_id(addr: &StacksAddress, name: &str) -> QualifiedContractIdentifier {
+    QualifiedContractIdentifier::new(StandardPrincipalData::from(addr.clone()), ContractName::try_from(name.to_string()).unwrap())
+}
+
+/// Extract a PoX address from its tuple representation
+fn tuple_to_pox_addr(tuple_data: TupleData) -> (AddressHashMode, Hash160) {
+    let version_value = tuple_data.get("version").expect("FATAL: no 'version' field in pox-addr").to_owned();
+    let hashbytes_value = tuple_data.get("hashbytes").expect("FATAL: no 'hashbytes' field in pox-addr").to_owned();
+
+    let version_u8 : u8 = version_value.expect_u128().try_into().expect("FATAL: PoX version is not a supported version byte");
+    let version : AddressHashMode = version_u8.try_into().expect("FATAL: PoX version is not a supported version byte");
+
+    let hashbytes_vec = hashbytes_value.expect_buff(20);
+
+    let mut hashbytes_20 = [0u8; 20];
+    hashbytes_20.copy_from_slice(&hashbytes_vec[0..20]);
+    let hashbytes = Hash160(hashbytes_20);
+
+    (version, hashbytes)
+}
+
+impl Value {
```

Comment:
Should we move this set of methods in `src/vm/types/mod.rs`? I'm pretty sure they would be re-used.

Reply:
Sure, can do :+1:

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/4e9abe5ac400e0597b12657325caf62ee4e170e5 / https://github.com/blockstack/stacks-blockchain/pull/1850/commits/cf799e56a66e0f8ab703d0619d568652dfacb6c8